### ISSUE TITLE
Update released version?

### ISSUE THREAD

**Issue by mattloose** [piccolomo/plotext#184]
Hi,

This is a great library - do you have plans to push an update to pypi soon supporting some of the recent changes? Other libraries building on this would benefit!

Thanks!

**Comment by piccolomo** [comment]
Hi apologies, I am working on version 6 by rewriting the code from scratch, and hopefully working on issue later. 

The code needs refreshing (especially in the _build.py file which is mysterious to read to me even)

**Comment by mattloose** [comment]
Thanks! 

I'm using the recent textual wrapper to plotext - this uses the pypi version which doesn't include the option to label stacked bar plots. 

It would be cool if it did in future!

**Comment by piccolomo** [comment]
Ok, thanks for replying.

I will try to work on labelled bar plots. For now the way to do it is 'manually' as shown [here](https://github.com/piccolomo/plotext/blob/master/readme/decorator.md#text-plot).

**Comment by blochl** [comment]
Italso would be awesome to have a released version with heatmap support...

**Comment by ap--** [comment]
Hi @piccolomo,

First, thanks for this great library ❤️ and sorry for bumping this issue. 
It would be wonderful if the current version on master could just be released as `5.2.9`.
If there are any open issues that would block a pre `v6` release, I would be happy to contribute time.

Cheers,
Andreas 😃

**Comment by piccolomo** [comment]
Hi @ap-- 

yes sorry you and others if waiting for version 6. I got derailed into rewriting the kernel of the program and it is taking more time then expected. 

You are welcomed to contribute. You could write me by email (piccolomo@gmail.com) and see how you could help.

Thanks for the message,
Savino

**Comment by piccolomo** [comment]
Just to update and apologise: I have been quite busy rewriting the package from ground up and will update.

**Comment by luxedo** [comment]
I discovered this package last week and it has been a lifesaver for my hobby project. Thank you!

I'd love to use the `box` in my project but it seems it has not been released yet at PyPI. If there's anything I can do to help release `5.3.1` please let me know.

**Comment by piccolomo** [comment]
Hi @luxedo,

so basically I have been working not stop (when I got the time) to build the project from ground up. Reason is: I was not happy with `plotext` kernel. I did not expect it to be this laborious! Dang it!

Here is the progress:
- the plotting kernel is ready! basically the plot central part, where data is plotted in colours and optionally with higher definition characters. This is the most difficult part

- I am now building the easy part: upper and lower bars, axis, numerical ticks

- Next difficult step is having subplots but I have the rough code rewritten already.

- From there I have to add all types of extra plots, like bar and box plot... 

It is going to take a while, especially cause I want to do it properly! 

I have lost memory of the state of `plotext` as it is now on GitHub. What I can do is publish it on [test pypi](https://test.pypi.org/project/plotext/) and give you the link so you can check if all works fine.

One you happy I can publish it on pypi. 

Would you be available? You would have to test the package. Would that work?

Thank you! 
Savino

**Comment by luxedo** [comment]
Yes I can take a look. I see there's a few tests in https://github.com/piccolomo/plotext/blob/master/tests.sh and it could be a start

**Comment by piccolomo** [comment]
Hi @luxedo,

here is version 5.3.1 on [testpypi](https://test.pypi.org/project/plotext/). Those tests are a good starting point.

Thank you!

**Comment by luxedo** [comment]
# Plotext 5.3.1 tests

> TLDR;
> Here are the outputs of the test script
> [test_report_split1.txt](https://github.com/user-attachments/files/17084544/test_report_split1.txt)
> [test_report_split2.txt](https://github.com/user-attachments/files/17084545/test_report_split2.txt)
> 
> It seems that everything is working except for the `YouTube` tests. Also, a coulple of plot's I'm working on were perfect.

## Installing
To install the package I had to add `--extra-index` for getting some dependencies from regular PyPI. Also, for some reason when I tried to install it was always in version in `test.pypy` is 4.2.0, so I had to specify 5.3.1.
```
pip install -i https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple 'plotext[video]==5.3.1'
``` 
## Running test command
Running the test script for the first time freezes the script in both *YouTube Plot* tests:
```bash
echo -en "Command Line Tool: YouTube Plot"
fork
if test $out -eq 1
then
    plotext youtube --url test
fi
# And
echo -en "YouTube Plot"
fork
if test $out -eq 1
then
    python3 -c "import plotext as plt; plt.play_youtube(plt.test_youtube_url)"
fi
```
Freezes here with the following message:
```
yes | ./tests.sh > test_report.txt

[ WARN:0@0.124] global cap.cpp:164 open VIDEOIO(CV_IMAGES): raised OpenCV exception:

OpenCV(4.10.0) /io/opencv/modules/videoio/src/cap_images.cpp:244: error: (-5:Bad argument) CAP_IMAGES: error, expected '0?[1-9][du]' pattern, got: https://rr2---sn-oxunxg8pjvn-jfcy.googlevideo.com/videoplayback?expire=1726952297&ei=Cd_uZqC8AZXb1sQP-KKouAw&ip=2804%3A14c%3A482%3A4f51%3A0%3A0%3A0%3A7990&id=o-AIQYfLTbAa_wKlrk4tKeSTGMD4VCuWm63Y87eHYNy6ed&itag=18&source=youtube&requiressl=yes&xpc=EgVo2aDSNQ%3D%3D&mh=tC&mm=31%2C29&mn=sn-oxunxg8pjvn-jfcy%2Csn-bg0s7nl6&ms=au%2Crdu&mv=m&mvi=2&pcm2cms=yes&pl=55&initcwndbps=1397500&bui=AXLXGFTrjvHJurxagIcmWjKf97paCVZtEVMjyjif9T1caYY5N6Tgo2NPiXqVTWTSyzdElN6xYEPLbUs5&spc=54MbxXgkMHtEV43WZ0DRiMIZsapeG8BFejY8Ki3VZpRh1Zvy0txQkeocciPTHVc&vprv=1&svpuc=1&mime=video%2Fmp4&ns=UfSMIpvi_QfYZqiGxCFPW8MQ&rqh=1&gir=yes&clen=59468543&ratebypass=yes&dur=1231.145&lmt=1698606074678367&mt=1726930383&fvip=4&c=WEB&sefc=1&txp=4438434&n=q-8e_0X-UewMwTP6A&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cxpc%2Cbui%2Cspc%2Cvprv%2Csvpuc%2Cmime%2Cns%2Crqh%2Cgir%2Cclen%2Cratebypass%2Cdur%2Clmt&sig=AJfQdSswRQIgM_l2a6GKYrCBCLnqJHgp225d6B4Oxqhhm5hGPmB8mX0CIQCcRX7A5aEx6obRCKFnH_b0t3PcrP2p8AtgYr5xe6TOGQ%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl%2Cinitcwndbps&lsig=ABPmVW0wRgIhAP02772Id0shTtbd_Bc3a-CbvWkJ9G5vfspfb9LRuu4UAiEAwWk-e6rSF0cFJ5b1jxeHCEv_wbv6GX0IV_PelV5B7J0%3D in function 'icvExtractPattern'


[https @ 0x7aac08001880] HTTP error 403 Forbidden
https://rr2---sn-oxunxg8pjvn-jfcy.googlevideo.com/videoplayback?expire=1726952297&ei=Cd_uZqC8AZXb1sQP-KKouAw&ip=2804%3A14c%3A482%3A4f51%3A0%3A0%3A0%3A7990&id=o-AIQYfLTbAa_wKlrk4tKeSTGMD4VCuWm63Y87eHYNy6ed&itag=18&source=youtube&requiressl=yes&xpc=EgVo2aDSNQ%3D%3D&mh=tC&mm=31%2C29&mn=sn-oxunxg8pjvn-jfcy%2Csn-bg0s7nl6&ms=au%2Crdu&mv=m&mvi=2&pcm2cms=yes&pl=55&initcwndbps=1397500&bui=AXLXGFTrjvHJurxagIcmWjKf97paCVZtEVMjyjif9T1caYY5N6Tgo2NPiXqVTWTSyzdElN6xYEPLbUs5&spc=54MbxXgkMHtEV43WZ0DRiMIZsapeG8BFejY8Ki3VZpRh1Zvy0txQkeocciPTHVc&vprv=1&svpuc=1&mime=video%2Fmp4&ns=UfSMIpvi_QfYZqiGxCFPW8MQ&rqh=1&gir=yes&clen=59468543&ratebypass=yes&dur=1231.145&lmt=1698606074678367&mt=1726930383&fvip=4&c=WEB&sefc=1&txp=4438434&n=q-8e_0X-UewMwTP6A&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cxpc%2Cbui%2Cspc%2Cvprv%2Csvpuc%2Cmime%2Cns%2Crqh%2Cgir%2Cclen%2Cratebypass%2Cdur%2Clmt&sig=AJfQdSswRQIgM_l2a6GKYrCBCLnqJHgp225d6B4Oxqhhm5hGPmB8mX0CIQCcRX7A5aEx6obRCKFnH_b0t3PcrP2p8AtgYr5xe6TOGQ%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpcm2cms%2Cpl%2Cinitcwndbps&lsig=ABPmVW0wRgIhAP02772Id0shTtbd_Bc3a-CbvWkJ9G5vfspfb9LRuu4UAiEAwWk-e6rSF0cFJ5b1jxeHCEv_wbv6GX0IV_PelV5B7J0%3D: Server returned 403 Forbidden (access denied)
```
The version `5.2.8` also prouces this error. After removing the `YouTube` tests, all the other seems to build a coherent plot.

**Comment by piccolomo** [comment]
I see : I am going to publish it regardless on pypi and deal with this video streaming issue for next version, as it is not central to the package.

Let me know if you or anyone else does not agree.

Thank you for the tests!

**Comment by piccolomo** [comment]
Published on [pypi](https://pypi.org/project/plotext/)! 

:-)

**Comment by Zaloog** [comment]
Unfortunately the new release is breaking [textual-plotext](https://github.com/Textualize/textual-plotext), a wrapper library to use plotext plots in [textual](https://textual.textualize.io), the actual code changes are not in github yet correct?

Best regards, and thanks for this awesome library and the work you put into it

**Comment by piccolomo** [comment]
Hi @Zaloog thanks for letting me know. I also just found out about your project which I found interesting!

I have tried this code from your readme, and it works just fine on Ubuntu. 

```
from textual.app import App, ComposeResult

from textual_plotext import PlotextPlot

class ScatterApp(App[None]):

    def compose(self) -> ComposeResult:
        yield PlotextPlot()

    def on_mount(self) -> None:
        plt = self.query_one(PlotextPlot).plt
        y = plt.sin() # sinusoidal test signal
        plt.scatter(y)
        plt.title("Scatter Plot") # to apply a title

if __name__ == "__main__":
    ScatterApp().run()
```

**Comment by Zaloog** [comment]
the specific error I get atm on the project im developing is `AttributeError: 'Plot' object has no attribute 'datetimes_to_string'`.

Also running the textual-plotext Demo with `python -m textual_plotext` fails on my machine (Windows)
``` 
                                                                                                                                                                                                                                                                                         │
│   466 │   │   ly = len(Y)                                                                      ╭────────────────────────────────────────────────────────────── locals ───────────────────────────────────────────────────────────────╮                                                   │
│   467 │   │   marker = [marker] * ly if marker is None or type(marker) != list else marker     │        args = (['Sausage', 'Pepperoni', 'Mushrooms', 'Cheese', 'Chicken', 'Beef'], [[14, 36, 11, 8, 7, 4], [12, 20, 35, 15, 2, 1]]) │                                                   │
│   468 │   │   color = [color] * ly if color is None else color                                 │       color = [None, None]                                                                                                          │                                                   │
│ ❱ 469 │   │   labels = [label] * ly if labels is None else labels                              │        fill = None                                                                                                                  │                                                   │
│   470 │   │   Y = ut.transpose([ut.cumsum(el) for el in ut.transpose(Y)])                      │      labels = None                                                                                                                  │                                                   │
│   471 │   │   for i in range(ly - 1, -1, -1):                                                  │          ly = 2                                                                                                                     │                                                   │
│   472 │   │   │   self.draw_bar(x, Y[i],                                                       │      marker = [None, None]                                                                                                          │                                                   │
│                                                                                                │     minimum = None                                                                                                                  │                                                   │
│                                                                                                │      offset = None                                                                                                                  │                                                   │
│                                                                                                │ orientation = None                                                                                                                  │                                                   │
│                                                                                                │ reset_ticks = None                                                                                                                  │                                                   │
│                                                                                                │        self = <plotext._monitor.monitor_class object at 0x000001FDD61F4650>                                                         │                                                   │
│                                                                                                │       width = None                                                                                                                  │                                                   │
│                                                                                                │           x = ['Sausage', 'Pepperoni', 'Mushrooms', 'Cheese', 'Chicken', 'Beef']                                                    │                                                   │
│                                                                                                │       xside = None                                                                                                                  │                                                   │
│                                                                                                │           Y = [[14, 36, 11, 8, 7, 4], [12, 20, 35, 15, 2, 1]]                                                                       │                                                   │
│                                                                                                │       yside = None                                                                                                                  │                                                   │
│                                                                                                ╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯                                                   │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
NameError: name 'label' is not defined
```
Both work with version `<5.3`, therefore I pinned the dependency to this currently.

**Comment by piccolomo** [comment]
Ok the `datetimes_to_string` function was updated to `datetimes_to_strings`; but I have added `datetimes_to_string` as alias in GitHub version. 

I have run your test again and I get the same error, which is very strange, because that method should be now available. Can you double check on your side?

Install GitHub version with `pip install git+https://github.com/piccolomo/plotext`.

**Comment by luxedo** [comment]
I can confirm that installing with `pip install git+https://github.com/piccolomo/plotext` works. And so does `5.2.8`.

**Comment by piccolomo** [comment]
Thanks @luxedo, did you get to test `python -m textual_plotext ` with github version?

**Comment by luxedo** [comment]
Yes! It works
```bash
$ uv pip install git+https://github.com/piccolomo/plotext
 Updated https://github.com/piccolomo/plotext (4d19108)
Resolved 1 package in 2.49s
   Built plotext @ git+https://github.com/piccolomo/plotext@4d19108b93e34a60ba789681756450ae126a76ed
Prepared 1 package in 659ms
Installed 1 package in 2ms
 + plotext==5.3.2 (from git+https://github.com/piccolomo/plotext@4d19108b93e34a60ba789681756450ae126a76ed)

$ uv run python -c "import plotext; print(plotext.__version__)"
warning: `uv run` is experimental and may change without warning.
5.3.2

$ uv run python -m textual_plotext  
```

![image](https://github.com/user-attachments/assets/cb7b61c2-6c05-49af-a544-5c9c49fc54c2)

**Comment by piccolomo** [comment]
Thanks, it does not work on my machine: I checked again. Anyway I will publish version 5.3.2 on pypi and delete 5.3.1, unless new urgent issues appear.

**Comment by piccolomo** [comment]
Version 5.3.2 published on [pypi](https://pypi.org/project/plotext/).

**Comment by Zaloog** [comment]
Also does not work on my machine with 
`plotext==5.3.2`
`textual-plotext==0.2.1`
Still getting `AttributeError: 'Plot' object has no attribute 'datetimes_to_string'`

In my project I just changed the function name to `datetimes_to_strings`, and the `label` argument to `labels`, which worked then with plotext 5.3.2.

**Comment by luxedo** [comment]
@Zaloog can you confirm that the version you're running is correct and there's not any other older version dangling in your system?

I tested in a clean environment and installed `plotext==5.3.2` and `textual-plotext==0.2.1` from PyPI. It worked given that this new version adds the alias for _datetimes_to_string**S**_  and also fixes the `label` issue: https://github.com/piccolomo/plotext/commit/a521629f14c14dabc8357041c101221afbef8f00

![image](https://github.com/user-attachments/assets/64795f03-dd3d-4c71-8fb8-061114007740)
![image](https://github.com/user-attachments/assets/bc0a04ab-d322-42d4-995d-81eb389acbb6)

The thing that you have to make sure is that the version `5.3.1` is not installed. You can try the following steps:
```bash
# Remove and then update textual-plotext
pip uninstall plotext
pip install -U textual-plotext

# Or you can try only updating plotext 
pip install -U plotext
```

If this doesn't work there's probably some issue with you python environments and libraries, which is loading an older version of `plotext`.

**Comment by Zaloog** [comment]
I resynched the environment before with a clear uv cache

![image](https://github.com/user-attachments/assets/ec5d7688-5d93-4ea6-b38d-011d966538eb)

**Comment by luxedo** [comment]
There's something very strange. I tried with my system's `pip` and still got it working. Steps to reproduce:

```bash
# Remove previous versions, purge cache and reinnstall
pip uninstall plotext --break-system-packages
pip cache purge
pip install plotext --break-system-packages

# Shows installation location
pip show plotext

# Installed version contains the working alias
rg "datetimes_to_string =" ~/.local/lib/python3.12/site-packages/plotext/_core.py
```

![image](https://github.com/user-attachments/assets/11416c2a-fa85-4899-a9fb-6b98e33e78a5)

Have you guys tried with `pip install --no-cache-dir plotext`?

**Comment by piccolomo** [comment]
Still not working. Also python uses an externally managed environment and it installs my packages in `~/pytorch_env/lib/python3.12/site-packages/plotext`.

**Comment by piccolomo** [comment]
Hi @luxedo, @Zaloog I removed some posts about `textual-plotext`, cause they were too many and a bit off topic. Long story short: it does not seem to work (or it works in some cases) and solution is not found yet. But I am still interested in a solution.

**Comment by duburcqa** [comment]
@piccolomo Thank you for this great library ! Would you mind to share with us a quick update of your progress ? I suggest that you open a dedicated issue about the process to easily keep track of it for anybody (as far as I know, it is possible to pin specific issues on top of the list to increase their visibility).

Edit: What about some dev branch for early-access testing by the community ?

**Comment by piccolomo** [comment]
Hi @duburcqa,

I’m glad to hear you’re enjoying the library, and I apologize for the delay. 

Here's the update: I’m currently working on the final part of a single simple plot, specifically adding the **legend**. 

This issue report is to one dedicated to updates, and I’ll try to pin this post along with a few others.

I’ve also created a separate and temporary private repository for version 6 to avoid any confusion with the current version. Do you think making it public would be helpful?

To @luxedo, I apologize for the delay and greatly appreciate your willingness to help. I hope you’re still available to contribute soon, as your ideas are highly valued.

I’ve been working intensely whenever possible, balancing work commitments, holidays, and a major relocation.

P.S. I have not found a way to pin posts. I just tried.

**Comment by duburcqa** [comment]
I would like to keep track of the progress, out of curiosity. So I don't want to make things more complicated for you! I list of checkbox with estimated date would be more than enough (if you have time for this of course!).

Here is how to pin an issue:

<img width="1076" alt="Image" src="https://github.com/user-attachments/assets/8c2a9f56-6844-421a-bdaa-6a4af5e6b1c7" />

**Comment by piccolomo** [comment]
Hi @duburcqa,

Apologies for the delay — it’s been a very busy period. Here’s a quick update:

- I’ve already built a simple plot.

- I’m currently working on the subplots. The core code is there, and I’m in the process of integrating everything.

It’s taking a bit longer than I expected, and I’m unsure how best to proceed after this stage. One option would be to publish a beta version in a secondary, temporary repository — though I’m slightly concerned it might cause confusion.

From there, I’ll continue following the timeline I’ve already outlined.

Thanks for your patience!

P.S. I have pinned the issue !

**Comment by mchav** [comment]
@piccolomo do you have a document outlining what the design will look like. I'm porting parts of this library to another language and was wondering how different a rewrite this will be.

**Comment by snooppr** [comment]
@piccolomo Hi, I want to thank you for this amazing lib. For me, the beauty of this lib is that it does not require heavy dependencies to be installed by default, for example, "numpy/pillow", i.e. it can be run even on Android, 

<details>
<summary> 🔵 for example</summary>

<img width="720" height="1440" alt="Image" src="https://github.com/user-attachments/assets/25d11a9c-7188-4a7d-921c-186a3a814e7c" />

Received stars of your repository, plotext in android/termux, I use in [shotstars](https://github.com/snooppr/shotstars).

</details>

Will Plotext remain as independent from third-party heavy libraries in version 6?

**Comment by pietrodantuono** [comment]
Hey @piccolomo, first of all thank you very much  for the effort. I would be happy to help with the new project if you want!

Pietro

**Comment by SteMazzO** [comment]
Any news @piccolomo ?

**Comment by piccolomo** [comment]
sorry every one.I am still working on it almost daily. I was scared to
watch the emails from Github but ye i own you an update. New version coming
soon. A little more of patience. Sorry this took way more then expected or
planned.

- Savino

**Comment by piccolomo** [comment]
P.S. I am basically now taking care of docstrings and public documentation.
Code wise it is done. I will give code a final read through to clean it and
improve it then i will publish on github first. Sort past  and new
unavoidable issues then pypi.

- Savino

**Comment by piccolomo** [comment]
Hi @mattloose, @blochl, @ap--, @luxedo, @Zaloog, @duburcqa, @mchav, @pietrodantuono, @SteMazzO, and everyone else who has been waiting here.

  **[plotext 6](https://plotext.readthedocs.io/en/latest/) is out, as a beta.**

  First, my apologies. This took years, far longer than I said it would, and for long stretches I went quiet, which was worse than the delay itself. I have been working on it almost daily, whenever life left me the time, some months a great deal and others very little; it never stopped, it only went slowly and quietly. Seeing now how much time it swallowed, I am frankly embarrassed, and I can only hope the rewritten kernel proves worth the wait.

  What happened is that a small clean-up turned into a rewrite of the whole package. 

  - the **drawing kernel is written in C++** and compiled when the package is installed, which is where most of those years went
  - the package is **reorganised from scratch**, one folder per concern, and a plot is now built on one master figure, `plotext.figure`, instead of package level functions
  - the **documentation is written from scratch** as a proper guide, and every method carries a colored docstring, browsable in the terminal
  - there is a **test suite** now, run with `plotext.test()`, and it runs on Linux, macOS and Windows on every change, none of which existed before
  - **Windows and macOS are supported and tested**; version 6 never compiled on either until a few days ago

  On the specific things asked in this thread, all of them are in:

  - @mattloose, **labels on bar plots**, stacked ones included, are a parameter now: see the [labeled bars](https://plotext.readthedocs.io/en/latest/bar.html#labeled-bars) section
  - @blochl, the **heatmap** is in, described [here](https://plotext.readthedocs.io/en/latest/specialized.html#heatmap)
  - @luxedo, the **box plot** is in and complete, described [here](https://plotext.readthedocs.io/en/latest/bar.html#box-plot); thank you again for running those 5.3.1 tests back then, they are what got that release out
  - @mchav, on the design: there is no separate design document, but the [guide](https://plotext.readthedocs.io/en/latest/) and the [change log](https://plotext.readthedocs.io/en/latest/changelog.html) together describe the whole shape of version 6 and every difference from version 5, which is what you would
  need for a port
  - @Zaloog, version 6 is a different interface, so [textual-plotext](https://github.com/Textualize/textual-plotext) will need work to sit on top of it; I am happy to help whoever takes that on
  - @ap--, @pietrodantuono, thank you both for offering to help, and sorry for not taking you up on it; the offer is worth much more now that there is a stable base to build on

  Where to find it:

  - GitHub: https://github.com/piccolomo/plotext
  - PyPI: https://pypi.org/project/plotext/6.0.0b0/, installed with `pip install --pre plotext`, the `--pre` being needed because it is a beta
  - documentation: https://plotext.readthedocs.io/en/latest/

  It is a beta because it is new, not because it is unfinished: what I want now is your reports. Try it, break it, and open an issue for anything that misbehaves or reads badly in the guide. I am going through every old issue and pull request one at a time, and answering them.

  Thank you all for the patience, and sorry again for how long it took,
  Savino
